opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java

```diff
@@ -191,33 +202,140 @@ private Optional<ExprType> type(String field) {
   }
 
   /**
-   * Only default strict_date_optional_time||epoch_millis is supported,
-   * strict_date_optional_time_nanos||epoch_millis if field is date_nanos.
-   * <a href="https://opensearch.org/docs/latest/opensearch/supported-field-types/date/#formats">
-   *   docs</a>
-   * The customized date_format is not supported.
+   * return the first matching formatter as an Instant to UTF.
+   *
+   * @param value - timestamp as string
+   * @param dateType - field type
+   * @return Instant without timezone
    */
-  private ExprValue constructTimestamp(String value) {
-    try {
-      return new ExprTimestampValue(
-          // Using OpenSearch DateFormatters for now.
-          DateFormatters.from(DATE_TIME_FORMATTER.parse(value)).toInstant());
-    } catch (DateTimeParseException e) {
-      throw new IllegalStateException(
+  private ExprValue parseTimestampString(String value, OpenSearchDateType dateType) {
+    Instant parsed = null;
+    for (DateFormatter formatter : dateType.getAllNamedFormatters()) {
+      try {
+        TemporalAccessor accessor = formatter.parse(value);
+        ZonedDateTime zonedDateTime = DateFormatters.from(accessor);
+        // remove the Zone
+        parsed = zonedDateTime.withZoneSameLocal(ZoneId.of("Z")).toInstant();
+      } catch (IllegalArgumentException ignored) {
+        // nothing to do, try another format
+      }
+    }
+
+    // FOLLOW-UP PR: Check custom formatters too
+
+    // if no named formatters are available, use the default
+    if (dateType.getAllNamedFormatters().size() == 0
+        || dateType.getAllCustomFormatters().size() > 0) {
+      try {
+        parsed = DateFormatters.from(DATE_TIME_FORMATTER.parse(value)).toInstant();
+      } catch (DateTimeParseException e) {
+        // ignored
+      }
+    }
+
+    if (parsed == null) {
+      // otherwise, throw an error that no formatters worked
+      throw new IllegalArgumentException(
           String.format(
-              "Construct ExprTimestampValue from \"%s\" failed, unsupported date format.", value),
-          e);
+              "Construct ExprTimestampValue from \"%s\" failed, unsupported date format.", value)
```

Comment:
Add IT with an unsupported format

Reply:
Unit tests are already added and may not need further testing with IT.